From 3951d8e0aa5100b8cc7c94f39a67d743f69f2a8b Mon Sep 17 00:00:00 2001
From: Rameshkumar Sundaram <quic_ramess@quicinc.com>
Date: Wed, 15 Jun 2022 08:29:43 +0530
Subject: [PATCH] wpa_supplicant: Add 5GHz 240MHz support for mesh

240MHz can be supported in 5 GHz for from channel
#100 - 144 with right hand 80MHz punctured in 320MHz bandwidth.
Add support to bring up mesh in 5 GHz 320MHz (240+80 Punctrued).
When 320MHz is configured for channel 100-144, static puncturing
pattern bits for higher 80MHz should be set (i.e. pattern & 0xF000)
should result in 0xF000).
EHT phy capability for advertising 320MHz support in 6GHz is reused
for 5GHz as well if 320MHz is configured.
Add the missing 80MHz channels to bw_80_160 array of
ibss_mesh_setup_freq().

Signed-off-by: Rameshkumar Sundaram <quic_ramess@quicinc.com>
---
 src/common/ieee802_11_common.c  |  3 ++-
 wpa_supplicant/ap.c             |  5 ++--
 wpa_supplicant/wpa_supplicant.c | 51 ++++++++++++++++++++++++++++-------------
 3 files changed, 40 insertions(+), 19 deletions(-)

--- a/src/common/ieee802_11_common.c
+++ b/src/common/ieee802_11_common.c
@@ -1028,7 +1028,8 @@ ieee80211_freq_to_channel_ext(unsigned i
 		if ((freq - 5000) % 5)
 			return NUM_HOSTAPD_MODES;
 
-		if (vht_opclass)
+		/* There is no op class defined for 5 GHz 320MHz, let it be unset */
+		if (vht_opclass || chanwidth == CONF_OPER_CHWIDTH_320MHZ)
 			*op_class = vht_opclass;
 		else if (sec_channel == 1)
 			*op_class = 122;
--- a/wpa_supplicant/ap.c
+++ b/wpa_supplicant/ap.c
@@ -105,9 +105,10 @@ static void wpas_conf_ap_vht(struct wpa_
 		hostapd_set_oper_centr_freq_seg0_idx(conf, freq_seg_idx);
 
 		wpa_printf(MSG_DEBUG,
-			   "VHT seg0 index %d and seg1 index %d for AP",
+			   "VHT seg0 index %d and seg1 index %d  BW %d for AP",
 			   hostapd_get_oper_centr_freq_seg0_idx(conf),
-			   hostapd_get_oper_centr_freq_seg1_idx(conf));
+			   hostapd_get_oper_centr_freq_seg1_idx(conf),
+			   hostapd_get_oper_chwidth(conf));
 		return;
 	}
